MqttSessionSpec: MQTT server connector should produce a duplicate publish on the server given two client connections

Appears to be flaky. Seen before #148 #46 

```
[info] MQTT server connector
[info] - should flow through a server session (68 milliseconds)
[info] - should receive two subscriptions for the same topic (38 milliseconds)
[info] - should unsubscribe a server session (131 milliseconds)
[info] - should reply to a ping request (38 milliseconds)
[info] - should close when no ping request received (1 second, 542 milliseconds)
[info] - should notify on disconnect (248 milliseconds)
[info] - should re-connect given connect, subscribe, disconnect, connect, publish (35 milliseconds)
[info] - should re-connect given connect, subscribe, connect again, publish (41 milliseconds)
[info] - should consume a duplicate publish on the server (27 milliseconds)
[info] - should produce a duplicate publish on the server given two client connections *** FAILED *** (15 seconds, 469 milliseconds)
[info]   java.lang.AssertionError: assertion failed: expected no StreamSupervisor children, but got [Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-320-0-ignoreSink#600515596], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-338-0-ignoreSink#392179906], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-327-0-unnamed#-375044653], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-328-0-unnamed#-28601930], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-337-0-unnamed#-269928412], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-329-0-unnamed#1296919811], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-319-0-unnamed#363979158], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-311-0-unnamed#-1039162337], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-309-0-unnamed#-702717299], Actor[pekko://mqtt-spec/system/Materializers/StreamSupervisor-4/flow-310-0-unnamed#233619764]]
[info]   at scala.Predef$.assert(Predef.scala:223)
[info]   at org.apache.pekko.stream.testkit.scaladsl.StreamTestKit$.$anonfun$assertNoChildren$2(StreamTestKit.scala:64)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.apache.pekko.testkit.TestKitBase.poll$2(TestKit.scala:342)
[info]   at org.apache.pekko.testkit.TestKitBase.awaitAssert(TestKit.scala:359)
[info]   at org.apache.pekko.testkit.TestKitBase.awaitAssert$(TestKit.scala:331)
[info]   at org.apache.pekko.testkit.TestKit.awaitAssert(TestKit.scala:982)
[info]   at org.apache.pekko.stream.testkit.scaladsl.StreamTestKit$.$anonfun$assertNoChildren$1(StreamTestKit.scala:61)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.apache.pekko.testkit.TestKitBase.within(TestKit.scala:429)
[info]   at org.apache.pekko.testkit.TestKitBase.within$(TestKit.scala:416)
[info]   at org.apache.pekko.testkit.TestKit.within(TestKit.scala:982)
[info]   at org.apache.pekko.testkit.TestKitBase.within(TestKit.scala:444)
[info]   at org.apache.pekko.testkit.TestKitBase.within$(TestKit.scala:444)
[info]   at org.apache.pekko.testkit.TestKit.within(TestKit.scala:982)
[info]   at org.apache.pekko.stream.testkit.scaladsl.StreamTestKit$.assertNoChildren(StreamTestKit.scala:61)
[info]   at org.apache.pekko.stream.testkit.scaladsl.StreamTestKit$.assertAllStagesStopped(StreamTestKit.scala:43)
[info]   at docs.scaladsl.MqttSessionSpec.$anonfun$new$155(MqttSessionSpec.scala:1921)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1240)
[info]   at org.apache.pekko.stream.connectors.testkit.scaladsl.LogCapturing.withFixture(LogCapturing.scala:70)
[info]   at org.apache.pekko.stream.connectors.testkit.scaladsl.LogCapturing.withFixture$(LogCapturing.scala:66)
[info]   at docs.scaladsl.MqttSessionSpec.withFixture(MqttSessionSpec.scala:45)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.invokeWithFixture$1(AnyWordSpecLike.scala:1238)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTest$1(AnyWordSpecLike.scala:1250)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTest(AnyWordSpecLike.scala:1250)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTest$(AnyWordSpecLike.scala:1232)
[info]   at docs.scaladsl.MqttSessionSpec.runTest(MqttSessionSpec.scala:45)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTests$1(AnyWordSpecLike.scala:1309)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:431)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:390)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:427)
[info]   at scala.collection.immutable.List.foreach(List.scala:431)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTests(AnyWordSpecLike.scala:1309)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.runTests$(AnyWordSpecLike.scala:1308)
[info]   at docs.scaladsl.MqttSessionSpec.runTests(MqttSessionSpec.scala:45)
[info]   at org.scalatest.Suite.run(Suite.scala:1114)
[info]   at org.scalatest.Suite.run$(Suite.scala:1096)
[info]   at docs.scaladsl.MqttSessionSpec.org$scalatest$wordspec$AnyWordSpecLike$$super$run(MqttSessionSpec.scala:45)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$run$1(AnyWordSpecLike.scala:1354)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.run(AnyWordSpecLike.scala:1354)
[info]   at org.scalatest.wordspec.AnyWordSpecLike.run$(AnyWordSpecLike.scala:1352)
[info]   at docs.scaladsl.MqttSessionSpec.org$scalatest$BeforeAndAfterAll$$super$run(MqttSessionSpec.scala:45)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at docs.scaladsl.MqttSessionSpec.run(MqttSessionSpec.scala:45)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.TestRunner.runTest$1(TestFramework.scala:153)
[info]   at sbt.TestRunner.run(TestFramework.scala:168)
[info]   at sbt.TestFramework$$anon$3$$anonfun$$lessinit$greater$1.$anonfun$apply$1(TestFramework.scala:336)
[info]   at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:296)
[info]   at sbt.TestFramework$$anon$3$$anonfun$$lessinit$greater$1.apply(TestFramework.scala:336)
[info]   at sbt.TestFramework$$anon$3$$anonfun$$lessinit$greater$1.apply(TestFramework.scala:336)
[info]   at sbt.TestFunction.apply(TestFramework.scala:348)
[info]   at sbt.Tests$.processRunnable$1(Tests.scala:475)
[info]   at sbt.Tests$.$anonfun$makeSerial$1(Tests.scala:481)
[info]   at sbt.std.Transform$$anon$3.$anonfun$apply$2(Transform.scala:47)
[info]   at sbt.std.Transform$$anon$4.work(Transform.scala:69)
[info]   at sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[info]   at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[info]   at sbt.Execute.work(Execute.scala:292)
[info]   at sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[info]   at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[info]   at sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:750)
```

--- a/mqtt-streaming/src/test/scala/docs/scaladsl/MqttSessionSpec.scala
+++ b/mqtt-streaming/src/test/scala/docs/scaladsl/MqttSessionSpec.scala
@@ -32,7 +32,7 @@ import pekko.testkit._
 import pekko.util.{ ByteString, Timeout }
 import org.scalatest.BeforeAndAfterAll
 import org.scalatest.concurrent.{ Eventually, ScalaFutures }
-import org.scalatest.time.{ Millis, Minutes, Span }
+import org.scalatest.time.{ Millis, Span }
 
 import scala.concurrent.{ ExecutionContext, Promise }
 import scala.concurrent.duration._
@@ -1922,159 +1922,156 @@ class MqttSessionSpec
       // longer patience needed since Akka 2.6
       implicit val patienceConfig: PatienceConfig = PatienceConfig(scaled(1.second), scaled(50.millis))
 
-      // https://github.com/apache/incubator-pekko-connectors/issues/148
-      eventually(timeout(Span(1, Minutes))) {
-        val serverSession = ActorMqttServerSession(settings.withProducerPubAckRecTimeout(10.millis))
+      val serverSession = ActorMqttServerSession(settings.withProducerPubAckRecTimeout(10.millis))
 
-        val client1 = TestProbe()
-        val toClient1 = Sink.foreach[ByteString](bytes => client1.ref ! bytes)
-        val (client1Connection, fromClient1) = Source
-          .queue[ByteString](1, OverflowStrategy.dropHead)
-          .toMat(BroadcastHub.sink)(Keep.both)
+      val client1 = TestProbe()
+      val toClient1 = Sink.foreach[ByteString](bytes => client1.ref ! bytes)
+      val (client1Connection, fromClient1) = Source
+        .queue[ByteString](1, OverflowStrategy.dropHead)
+        .toMat(BroadcastHub.sink)(Keep.both)
+        .run()
+
+      val pipeToClient1 = Flow.fromSinkAndSource(toClient1, fromClient1)
+
+      val client2 = TestProbe()
+      val toClient2 = Sink.foreach[ByteString](bytes => client2.ref ! bytes)
+      val (client2Connection, fromClient2) = Source
+        .queue[ByteString](0, OverflowStrategy.dropHead)
+        .toMat(BroadcastHub.sink)(Keep.both)
+        .run()
+
+      val pipeToClient2 = Flow.fromSinkAndSource(toClient2, fromClient2)
+
+      val clientId = "some-client-id"
+
+      val connect = Connect(clientId, ConnectFlags.None)
+      val connect1Received = Promise[Done]()
+      val connect2Received = Promise[Done]()
+
+      val subscribe = Subscribe("some-topic")
+      val subscribe1Received = Promise[Done]()
+      val subscribe2Received = Promise[Done]()
+
+      val pubAckReceived = Promise[Done]()
+
+      val disconnect = Disconnect
+      val disconnectReceived = Promise[Done]()
+
+      val serverConnection1 =
+        Source
+          .queue[Command[Nothing]](1, OverflowStrategy.fail)
+          .via(
+            Mqtt
+              .serverSessionFlow(serverSession, ByteString("connection 1"))
+              .join(pipeToClient1))
+          .wireTap(Sink.foreach[Either[DecodeError, Event[_]]] {
+            case Right(Event(`connect`, _)) =>
+              connect1Received.success(Done)
+            case Right(Event(cp: Subscribe, _)) if cp.topicFilters == subscribe.topicFilters =>
+              subscribe1Received.success(Done)
+            case Right(Event(`disconnect`, _)) =>
+              disconnectReceived.success(Done)
+            case other => fail(s"didn't match `$other`")
+          })
+          .toMat(Sink.seq)(Keep.left)
           .run()
 
-        val pipeToClient1 = Flow.fromSinkAndSource(toClient1, fromClient1)
+      val connectBytes = connect.encode(ByteString.newBuilder).result()
+      val connAck = ConnAck(ConnAckFlags.None, ConnAckReturnCode.ConnectionAccepted)
+      val connAckBytes = connAck.encode(ByteString.newBuilder).result()
 
-        val client2 = TestProbe()
-        val toClient2 = Sink.foreach[ByteString](bytes => client2.ref ! bytes)
-        val (client2Connection, fromClient2) = Source
-          .queue[ByteString](0, OverflowStrategy.dropHead)
-          .toMat(BroadcastHub.sink)(Keep.both)
+      val subscribeBytes = subscribe.encode(ByteString.newBuilder, PacketId(1)).result()
+      val subAck = SubAck(PacketId(1), List(ControlPacketFlags.QoSAtLeastOnceDelivery))
+      val subAckBytes = subAck.encode(ByteString.newBuilder).result()
+
+      val publish = Publish("some-topic", ByteString("some-payload"))
+      val publishBytes = publish.encode(ByteString.newBuilder, Some(PacketId(1))).result()
+      val dupPublishBytes = publish
+        .copy(flags = publish.flags | ControlPacketFlags.DUP)
+        .encode(ByteString.newBuilder, Some(PacketId(1)))
+        .result()
+      val pubAck = PubAck(PacketId(1))
+      val pubAckBytes = pubAck.encode(ByteString.newBuilder).result()
+
+      val disconnectBytes = disconnect.encode(ByteString.newBuilder).result()
+
+      client1Connection.offer(connectBytes)
+
+      connect1Received.future.futureValue shouldBe Done
+
+      serverConnection1.offer(Command(connAck))
+      client1.expectMsg(connAckBytes)
+
+      client1Connection.offer(subscribeBytes)
+
+      subscribe1Received.future.futureValue shouldBe Done
+
+      serverConnection1.offer(Command(subAck))
+      client1.expectMsg(subAckBytes)
+
+      serverSession ! Command(publish)
+      client1.expectMsg(publishBytes)
+
+      // Perform an explicit disconnect otherwise, if for example, we
+      // just completed the client connection, the session may receive
+      // the associated ConnectionLost signal for the new connection
+      // given that the new connection occurs so quickly.
+      client1Connection.offer(disconnectBytes)
+
+      disconnectReceived.future.futureValue shouldBe Done
+
+      val serverConnection2 =
+        Source
+          .queue[Command[Nothing]](1, OverflowStrategy.fail)
+          .via(
+            Mqtt
+              .serverSessionFlow(serverSession, ByteString("connection 2"))
+              .join(pipeToClient2))
+          .wireTap(Sink.foreach[Either[DecodeError, Event[_]]] {
+            case Right(Event(`connect`, _)) =>
+              connect2Received.success(Done)
+            case Right(Event(cp: Subscribe, _)) if cp.topicFilters == subscribe.topicFilters =>
+              subscribe2Received.success(Done)
+            case Right(Event(_: PubAck, _)) =>
+              pubAckReceived.success(Done)
+            case other => fail(s"didn't match `$other`")
+          })
+          .toMat(Sink.seq)(Keep.left)
           .run()
 
-        val pipeToClient2 = Flow.fromSinkAndSource(toClient2, fromClient2)
+      client2Connection.offer(connectBytes)
 
-        val clientId = "some-client-id"
+      connect2Received.future.futureValue shouldBe Done
 
-        val connect = Connect(clientId, ConnectFlags.None)
-        val connect1Received = Promise[Done]()
-        val connect2Received = Promise[Done]()
+      serverConnection2.offer(Command(connAck))
+      client2.expectMsg(6.seconds, connAckBytes)
 
-        val subscribe = Subscribe("some-topic")
-        val subscribe1Received = Promise[Done]()
-        val subscribe2Received = Promise[Done]()
+      client2Connection.offer(subscribeBytes)
 
-        val pubAckReceived = Promise[Done]()
+      subscribe2Received.future.futureValue shouldBe Done
 
-        val disconnect = Disconnect
-        val disconnectReceived = Promise[Done]()
+      serverConnection2.offer(Command(subAck))
 
-        val serverConnection1 =
-          Source
-            .queue[Command[Nothing]](1, OverflowStrategy.fail)
-            .via(
-              Mqtt
-                .serverSessionFlow(serverSession, ByteString.empty)
-                .join(pipeToClient1))
-            .wireTap(Sink.foreach[Either[DecodeError, Event[_]]] {
-              case Right(Event(`connect`, _)) =>
-                connect1Received.success(Done)
-              case Right(Event(cp: Subscribe, _)) if cp.topicFilters == subscribe.topicFilters =>
-                subscribe1Received.success(Done)
-              case Right(Event(`disconnect`, _)) =>
-                disconnectReceived.success(Done)
-              case other => fail(s"didn't match `$other`")
-            })
-            .toMat(Sink.seq)(Keep.left)
-            .run()
-
-        val connectBytes = connect.encode(ByteString.newBuilder).result()
-        val connAck = ConnAck(ConnAckFlags.None, ConnAckReturnCode.ConnectionAccepted)
-        val connAckBytes = connAck.encode(ByteString.newBuilder).result()
-
-        val subscribeBytes = subscribe.encode(ByteString.newBuilder, PacketId(1)).result()
-        val subAck = SubAck(PacketId(1), List(ControlPacketFlags.QoSAtLeastOnceDelivery))
-        val subAckBytes = subAck.encode(ByteString.newBuilder).result()
-
-        val publish = Publish("some-topic", ByteString("some-payload"))
-        val publishBytes = publish.encode(ByteString.newBuilder, Some(PacketId(1))).result()
-        val dupPublishBytes = publish
-          .copy(flags = publish.flags | ControlPacketFlags.DUP)
-          .encode(ByteString.newBuilder, Some(PacketId(1)))
-          .result()
-        val pubAck = PubAck(PacketId(1))
-        val pubAckBytes = pubAck.encode(ByteString.newBuilder).result()
-
-        val disconnectBytes = disconnect.encode(ByteString.newBuilder).result()
-
-        client1Connection.offer(connectBytes)
-
-        connect1Received.future.futureValue shouldBe Done
-
-        serverConnection1.offer(Command(connAck))
-        client1.expectMsg(connAckBytes)
-
-        client1Connection.offer(subscribeBytes)
-
-        subscribe1Received.future.futureValue shouldBe Done
-
-        serverConnection1.offer(Command(subAck))
-        client1.expectMsg(subAckBytes)
-
-        serverSession ! Command(publish)
-        client1.expectMsg(publishBytes)
-
-        // Perform an explicit disconnect otherwise, if for example, we
-        // just completed the client connection, the session may receive
-        // the associated ConnectionLost signal for the new connection
-        // given that the new connection occurs so quickly.
-        client1Connection.offer(disconnectBytes)
-
-        disconnectReceived.future.futureValue shouldBe Done
-
-        val serverConnection2 =
-          Source
-            .queue[Command[Nothing]](1, OverflowStrategy.fail)
-            .via(
-              Mqtt
-                .serverSessionFlow(serverSession, ByteString.empty)
-                .join(pipeToClient2))
-            .wireTap(Sink.foreach[Either[DecodeError, Event[_]]] {
-              case Right(Event(`connect`, _)) =>
-                connect2Received.success(Done)
-              case Right(Event(cp: Subscribe, _)) if cp.topicFilters == subscribe.topicFilters =>
-                subscribe2Received.success(Done)
-              case Right(Event(_: PubAck, _)) =>
-                pubAckReceived.success(Done)
-              case other => fail(s"didn't match `$other`")
-            })
-            .toMat(Sink.seq)(Keep.left)
-            .run()
-
-        client2Connection.offer(connectBytes)
-
-        connect2Received.future.futureValue shouldBe Done
-
-        serverConnection2.offer(Command(connAck))
-        client2.expectMsg(6.seconds, connAckBytes)
-
-        client2Connection.offer(subscribeBytes)
-
-        subscribe2Received.future.futureValue shouldBe Done
-
-        serverConnection2.offer(Command(subAck))
-
-        client2.fishForMessage(3.seconds.dilated) {
-          case msg: ByteString if msg == dupPublishBytes => true
-          case _                                         => false
-        }
-
-        client2Connection.offer(pubAckBytes)
-        pubAckReceived.future.futureValue shouldBe Done
-
-        client1Connection.complete()
-        client2Connection.complete()
-        serverConnection1.complete()
-        serverConnection2.complete()
-
-        for {
-          _ <- client1Connection.watchCompletion()
-          _ <- client2Connection.watchCompletion()
-          _ <- serverConnection1.watchCompletion()
-          _ <- serverConnection2.watchCompletion()
-        } serverSession.shutdown()
-
+      client2.fishForMessage(3.seconds.dilated) {
+        case msg: ByteString if msg == dupPublishBytes => true
+        case _                                         => false
       }
+
+      client2Connection.offer(pubAckBytes)
+      pubAckReceived.future.futureValue shouldBe Done
+
+      client1Connection.complete()
+      client2Connection.complete()
+      serverConnection1.complete()
+      serverConnection2.complete()
+
+      for {
+        _ <- client1Connection.watchCompletion()
+        _ <- client2Connection.watchCompletion()
+        _ <- serverConnection1.watchCompletion()
+        _ <- serverConnection2.watchCompletion()
+      } serverSession.shutdown()
+
     }
   }
 

